### PR TITLE
Fix extension size increase

### PR DESCRIPTION
**Pull request by samansmink**
In v1.1.0, the addition of the C Extension API has caused the extension binaries to increase quite significantly:

name | v1.0.0 uncompressed | v1.1.0 uncompressed
-- | -- | --
json | 29M | 77M
spatial | 55M | 110M
sqlite | 27M | 77M
prql | 18M | 40M

The reason is that due to the way we build extensions, we statically link duckdb in its entirety into the extension binary, then rely on the linker being able to prune the unused parts. What this means is that its quite easy unknowingly link in significant parts of DuckDB causing the extension binary to significantly grow.

This is precisely what happened when someone (possibly me) wrote some code that produces a struct containing all C API function pointers. Since the C API touches a lot of things throughout DuckDB this prevented the linker from being able to prune much.

The solution is to generate the struct of function pointers centrally in the DatabaseInstance, then hand out pointers to those.

@Mytherin I think we have 2 options: either put this behind a lock and lazily initialize, or initialize this in `DatabaseInstance::Initialize` then pass const pointers, I went with the latter.

Also I added a regression test for this that tests the size of `[json,parquet,tpch,tpcds]`. (theshold for check failure is currently at 20%).

### CODE REVIEW

File: .github/workflows/Regression.yml [hunk 1]
```diff
@@ -211,6 +211,50 @@ jobs:
...
+  name: Regression test binary size
+  runs-on: ubuntu-20.04
+  env:
+    CC: gcc-10
+    CXX: g++-10
+    GEN: ninja
+    BUILD_TPCH: 1
+    BUILD_TPCDS: 1
+    BUILD_JSON: 1
+    BUILD_PARQUET: 1
```
Comment (carlopi) [inline .github/workflows/Regression.yml:224]:
```suggestion
    BUILD_PARQUET: 1
    EXTENSION_STATIC_BUILD: 1
```
I think this is needed to actually show differences in a model that is closer to regular deployment of extensions

File: src/main/database.cpp [hunk 2]
```diff
@@ -272,6 +272,8 @@ void DatabaseInstance::Initialize(const char *database_path, DBConfig *user_conf
 		config.options.temporary_directory = string();
 	}
 
+	extension_api_v0 = make_uniq<duckdb_ext_api_v0>(CreateAPIv0());
+
```
Comment (carlopi) [inline src/main/database.cpp:276]: Just a note: `CreateAPIv0` can be tagged as constexpr starting in C++14, (not in C++11 due to more restrictive rules on assigments within constexpr) so I would assume most compilers to compile this away to just a bunch of memsets.

**Review by carlopi — APPROVED** [review]
Left a note on improving the test (that currently is 1-1, while in this PR should still be significant, see https://github.com/duckdb/duckdb/actions/runs/11131102326/job/30931920185?pr=14185), and another one on the fact that GetAPIV0 is basically constexpr, so I would assume basically has no cost (and we expect you to use it)

Thanks!!

**Review by carlopi — CHANGES_REQUESTED** [review]
Actually, it's a "request changes", due to the test that is not super informative otherwise

**Review by carlopi — APPROVED** [review]

**Comment by Mytherin** [comment]
Thanks for the PR! Great find.

I'm guessing the issue here is that `CreateAPIv0` is called in a static method (`GetAPI`) - when elsewhere we get a function pointer to that static method (in `CreateAccessStruct`) - likely preventing that call from being optimized out - causing the entire C API to always be statically linked in/never pruned.

Given that this is the problem - I wonder if we can simplify this solution. Do we need to hand out pointers to the struct that resides in the `DatabaseInstance`, or could we also just make a copy? That would avoid the need for locking and avoids introducing additional unnecessary lifetime problems.

**Comment by samansmink** [comment]
@Mytherin do you mean like this? ([7edb53d](https://github.com/duckdb/duckdb/pull/14185/commits/7edb53dd9ce1d23d21ac5aa9f3f6cd9c9ab3aa3c))

I think the convention should be that extensions copy the struct over into some global variable anyway, at least thats what the macro's in the duckdb_extensions.h do anyway. However I could see how leaving the pointers valid could results in people incorrectly assuming they are **always** valid and lifetime issues arising. To resolve that I have made an **extra** copy that ensures the pointer to the function pointer struct that the extension sees during initialization will only be valid during the initialization forcing the extension to copy it.

**Comment by carlopi** [comment]
@Mytherin: I think this is ready

Results (as per the added check) are:
```
- checking 'tpch': old size=41925566, new_size=14761342
 - checking 'json': old size=41906006, new_size=34334838
 - checking 'parquet': old size=41953918, new_size= 36586230
 - checking 'tpcds': old size=41962926, new_size=16051550
```

**Comment by samansmink** [comment]
1 sec, I will try to refactor it so that we don't need to store the struct in the databaseinstance. 

I think either through using a virtual method or storing the function pointer to the `CreateAPIv0()`  in the databaseinstance that **should** work

**Comment by samansmink** [comment]
@Mytherin Tried making a virtual `DatabaseInstance::GetExtensionAPIV0()` method, but that resulting in some weird crashes on loading extensions.

Went with the function pointer approach, which seems clean enough to me

**Comment by Mytherin** [comment]
That's cleaner - but I wonder if this does not also work? In that case we would not need to store anything in the database instance.

```cpp

const duckdb_ext_api_v0 DatabaseInstance::GetExtensionAPIV0() {
	return CreateAPIv0();
}

```

**Comment by samansmink** [comment]
No that will get the size right back up where it was

**Comment by Mytherin** [comment]
Alright, let's get this in as-is then. Thanks!

**Comment by samansmink** [comment]
Let's do a final double check for the binary sizes before merge. I checked locally, but better safe than sorry

**Comment by samansmink** [comment]
sizes seem all good:
```
 - checking 'tpch': old size=41913902, new_size=14765854
 - checking 'json': old size=41894342, new_size=34335206
 - checking 'parquet': old size=41942070, new_size=36578406
 - checking 'tpcds': old size=41951254, new_size=16051966
 ```

**Pull request merged**